src/core/src/AssetRegistry.cpp

```diff
@@ -1,108 +1,286 @@
-    _ionImageries.insert(_ionImageries.end(), std::make_shared<OmniIonImagery>(path));
+OmniIonImagery& AssetRegistry::addIonImagery(const PXR_NS::SdfPath& path) {
+    return *_ionImageries.insert(_ionImageries.end(), std::make_unique<OmniIonImagery>(_pContext, path))->get();
 }
 
-void AssetRegistry::addPolygonImagery(const pxr::SdfPath& path) {
-    _polygonImageries.insert(_polygonImageries.end(), std::make_shared<OmniPolygonImagery>(path));
+void AssetRegistry::removeIonImagery(const PXR_NS::SdfPath& path) {
+    CppUtil::eraseIf(_ionImageries, [&path](const auto& pIonImagery) { return pIonImagery->getPath() == path; });
 }
 
-void AssetRegistry::removeIonImagery(const pxr::SdfPath& path) {
-    _ionImageries.remove_if([&path](const auto& imagery) { return imagery->getPath() == path; });
+OmniIonImagery* AssetRegistry::getIonImagery(const PXR_NS::SdfPath& path) const {
+    for (const auto& pIonImagery : _ionImageries) {
+        if (pIonImagery->getPath() == path) {
+            return pIonImagery.get();
+        }
+    }
+
+    return nullptr;
+}
+
+OmniIonImagery* AssetRegistry::getIonImageryByIonAssetId(int64_t ionAssetId) const {
+    for (const auto& pIonImagery : _ionImageries) {
+        if (pIonImagery->getIonAssetId() == ionAssetId) {
+            return pIonImagery.get();
+        }
+    }
+
+    return nullptr;
+}
+
+const std::vector<std::unique_ptr<OmniIonImagery>>& AssetRegistry::getAllIonImageries() const {
+    return _ionImageries;
 }
 
-void AssetRegistry::removePolygonImagery(const pxr::SdfPath& path) {
-    _polygonImageries.remove_if([&path](const auto& imagery) { return imagery->getPath() == path; });
+OmniPolygonImagery& AssetRegistry::addPolygonImagery(const PXR_NS::SdfPath& path) {
+    return *_polygonImageries.insert(_polygonImageries.end(), std::make_unique<OmniPolygonImagery>(_pContext, path))
+                ->get();
 }
 
-std::shared_ptr<OmniIonImagery> AssetRegistry::getIonImagery(const pxr::SdfPath& path) const {
-    for (const auto& imagery : _ionImageries) {
-        if (imagery->getPath() == path) {
-            return imagery;
+void AssetRegistry::removePolygonImagery(const PXR_NS::SdfPath& path) {
+    CppUtil::eraseIf(
+        _polygonImageries, [&path](const auto& pPolygonImagery) { return pPolygonImagery->getPath() == path; });
+}
+
+OmniPolygonImagery* AssetRegistry::getPolygonImagery(const PXR_NS::SdfPath& path) const {
+    for (const auto& pPolygonImagery : _polygonImageries) {
+        if (pPolygonImagery->getPath() == path) {
+            return pPolygonImagery.get();
         }
     }
 
     return nullptr;
 }
 
-std::shared_ptr<OmniPolygonImagery> AssetRegistry::getPolygonImagery(const pxr::SdfPath& path) const {
-    for (const auto& imagery : _polygonImageries) {
-        if (imagery->getPath() == path) {
-            return imagery;
+const std::vector<std::unique_ptr<OmniPolygonImagery>>& AssetRegistry::getAllPolygonImageries() const {
```

Comment:
Kind of a nit and coming from C# where properties/getter/setter names are strictly defined, I wonder if it makes sense to drop change this to `getPolygonImageries()`.  The `All` seems redundant if the names is plural and returns a container. (I can swap this in my PR--no need here.)

Reply:
Good point, and I can make the change here.